File: App/Resources/Algorithms/lsdradixsort.js
// Aux arrays are write-only from the engine's perspective (matching RecordingEngine's read-via-
// tape-only design), so counts/values are tracked in plain JS arrays alongside the recorded
// writeAux calls used purely for visualization.
function sort(engine) {
    const n = engine.count();
    const radix = 4;

    function getDigit(value, place) {
        return Math.floor(value / Math.pow(radix, place)) % radix;
    }

    let maxValue = 0;
    for (let i = 0; i < n; i++) {
        maxValue = Math.max(maxValue, engine.getValue(i));
    }
    let highestPlace = 1;
    while (Math.pow(radix, highestPlace) <= maxValue) {
        highestPlace++;
    }

    const outputHandle = engine.createAuxArray(n);

    for (let place = 0; place < highestPlace; place++) {
        const counts = new Array(radix).fill(0);
        const values = [];
        for (let i = 0; i < n; i++) {
            values.push(engine.getValue(i));
        }
        for (let i = 0; i < n; i++) {
            counts[getDigit(values[i], place)]++;
        }
        for (let d = 1; d < radix; d++) {
            counts[d] += counts[d - 1];
        }
        const output = new Array(n);
        for (let i = n - 1; i >= 0; i--) {
            const digit = getDigit(values[i], place);
            counts[digit]--;
            output[counts[digit]] = values[i];
            engine.writeAux(outputHandle, counts[digit], values[i]);
        }
        for (let i = 0; i < n; i++) {
            engine.setValue(i, output[i]);
        }
    }

    engine.deleteAuxArray(outputHandle);
}

File: Modules/BuiltInAlgorithms/Sources/Distribution/LSDRadixSort.swift
import AlgorithmKit
import Foundation
import SortEngineKit

public struct LSDRadixSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "lsdradixsort")
  public let metadata = AlgorithmMetadata(
    displayName: "LSD Radix Sort",
    category: .distribution,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 4292, coefficients: [239924, 105.28, 0.0115046],
      measuredSafeCeiling: nil),
    detectedGrowthModel: DetectedGrowthModel(
      family: .polynomialIntercept, coefficients: [0.0115046, 6.52513, -11.2883], rSquared: 0.999346),
    stable: true,
    timeComplexity: ComplexityBounds(
      best: "O(d \\times (n+b))", average: "O(d \\times (n+b))", worst: "O(d \\times (n+b))"
    ),
    spaceComplexity: "O(n + b)",
    iconName: "number"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    let radix = 4

    func getDigit(_ value: Int, _ place: Int) -> Int {
      (value / Int(pow(Double(radix), Double(place)))) % radix
    }

    var maxValue = 0
    for i in 0..<n {
      maxValue = max(maxValue, engine.values[i])
    }
    var highestPlace = 1
    while Int(pow(Double(radix), Double(highestPlace))) <= maxValue {
      highestPlace += 1
    }

    let outputHandle = engine.createAuxArray(length: n)
    // Reused across every pass instead of allocated fresh each time -- both are fully
    // overwritten by the end of each pass (`values` by the snapshot loop below, `output` by the
    // partitioning loop, since `counts`' prefix sum accounts for every index exactly once), so
    // there's no stale-data risk in keeping the same backing storage across passes.
    var values = [Int](repeating: 0, count: n)
    var output = [Int](repeating: 0, count: n)

    for place in 0..<highestPlace {
      var counts = [Int](repeating: 0, count: radix)
      for i in 0..<n {
        values[i] = engine.values[i]
      }
      for i in 0..<n {
        counts[getDigit(values[i], place)] += 1
      }
      for d in 1..<radix {
        counts[d] += counts[d - 1]
      }
      for i in stride(from: n - 1, through: 0, by: -1) {
        let digit = getDigit(values[i], place)
        counts[digit] -= 1
        output[counts[digit]] = values[i]
        engine.writeAux(outputHandle, at: counts[digit], value: values[i])
      }
      for i in 0..<n {
        engine.setValue(i, output[i])
      }
    }

    engine.deleteAuxArray(outputHandle)
  }
}
